Add some investigation code for dumping additional info when MarkedBlock::Handle is invalid

danlliu opened the pull request:
#### fdd1cb2d251454b026727ba36e3a0c12cc0f08c5
<pre>
Add some investigation code for dumping additional info when MarkedBlock::Handle is invalid
<a href="https://bugs.webkit.org/show_bug.cgi?id=280370">https://bugs.webkit.org/show_bug.cgi?id=280370</a>
<a href="https://rdar.apple.com/136782494">rdar://136782494</a>

Reviewed by Mark Lam and Keith Miller.

We want to dump additional information when this failure case is hit to understand why this
could be taking place.

This patch is an updated version of Yijia&apos;s patch, which implemented most of the logging.

* Source/JavaScriptCore/heap/BlockDirectory.cpp:
(JSC::BlockDirectory::findMarkedBlockHandle):
* Source/JavaScriptCore/heap/BlockDirectory.h:
* Source/JavaScriptCore/heap/CellContainerInlines.h:
(JSC::CellContainer::aboutToMark):
* Source/JavaScriptCore/heap/HeapInlines.h:
(JSC::Heap::testAndSetMarked):
* Source/JavaScriptCore/heap/MarkedBlock.cpp:
(JSC::MarkedBlock::aboutToMarkSlow):
(JSC::MarkedBlock::dumpInfoIfHandleIsNotValid):
* Source/JavaScriptCore/heap/MarkedBlock.h:
(JSC::MarkedBlock::Handle::hasBlock):
(JSC::MarkedBlock::Header::handleBitsForNullCheck):
(JSC::MarkedBlock::aboutToMark):
* Source/JavaScriptCore/heap/MarkedSpace.cpp:
(JSC::MarkedSpace::findMarkedBlockHandle):
* Source/JavaScriptCore/heap/MarkedSpace.h:
* Source/JavaScriptCore/heap/SlotVisitorInlines.h:
(JSC::SlotVisitor::appendUnbarriered):
(JSC::SlotVisitor::appendHiddenUnbarriered):

Canonical link: <a href="https://commits.webkit.org/284376@main">https://commits.webkit.org/284376@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/ae130af84627e634b88fe272a53cc545411b5d98

| Misc | iOS, visionOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/69194 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/48/builds/48594 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/55/builds/21866 "Built successfully") | [  ~~🛠 wpe~~](https://ews-build.webkit.org/#/builders/5/builds/73275 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🛠 win~~](https://ews-build.webkit.org/#/builders/59/builds/20352 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") 
| | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/49/builds/56395 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/61/builds/20201 "Built successfully") | [  ~~🧪 wpe-wk2~~](https://ews-build.webkit.org/#/builders/5/builds/73275 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 win-tests~~](https://ews-build.webkit.org/#/builders/59/builds/20352 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/72260 "Passed tests") | [  ~~🧪 ios-wk2~~](https://ews-build.webkit.org/#/builders/47/builds/44351 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 api-mac~~](https://ews-build.webkit.org/#/builders/18/builds/59739 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 api-wpe~~](https://ews-build.webkit.org/#/builders/5/builds/73275 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/42/builds/41020 "Passed tests") | [  ~~🧪 mac-wk1~~](https://ews-build.webkit.org/#/builders/64/builds/17169 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🛠 wpe-cairo~~](https://ews-build.webkit.org/#/builders/65/builds/18726 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | 
| [  ~~🛠 🧪 jsc~~](https://ews-build.webkit.org/#/builders/20/builds/62310 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 api-ios~~](https://ews-build.webkit.org/#/builders/13/builds/62963 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 mac-wk2~~](https://ews-build.webkit.org/#/builders/63/builds/17514 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🛠 gtk~~](https://ews-build.webkit.org/#/builders/2/builds/74986 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | 
| [✅ 🛠 🧪 jsc-arm64](https://ews-build.webkit.org/#/builders/12/builds/68440 "Built successfully and passed tests") | [✅ 🛠 vision](https://ews-build.webkit.org/#/builders/87/builds/13176 "Built successfully") | [  ~~🧪 mac-AS-debug-wk2~~](https://ews-build.webkit.org/#/builders/62/builds/16750 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 gtk-wk2~~](https://ews-build.webkit.org/#/builders/2/builds/74986 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | 
| | [✅ 🛠 vision-sim](https://ews-build.webkit.org/#/builders/86/builds/13215 "Built successfully") | [  ~~🧪 mac-wk2-stress~~](https://ews-build.webkit.org/#/builders/8/builds/59822 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 api-gtk~~](https://ews-build.webkit.org/#/builders/2/builds/74986 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | 
| | [  ~~🧪 vision-wk2~~](https://ews-build.webkit.org/#/builders/88/builds/10631 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [⏳ 🧪 mac-intel-wk2](https://ews-build.webkit.org/#/builders/macOS-Ventura-Release-WK2-Intel-Tests-EWS "Waiting to run tests") | [✅ 🛠 jsc-armv7](https://ews-build.webkit.org/#/builders/35/builds/90221 "Built successfully") | 
| [✅ 🛠 🧪 unsafe-merge](https://ews-build.webkit.org/#/builders/22/builds/10567 "Built successfully and passed tests") | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/44/builds/44398 "Built successfully") | | [  ~~🧪 jsc-armv7-tests~~](https://ews-build.webkit.org/#/builders/25/builds/16004 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/46/builds/45471 "Built successfully") | | | 
| | [  ~~🛠 watch~~](https://ews-build.webkit.org/#/builders/43/builds/46667 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/45/builds/45213 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

kmiller68 commented on Source/JavaScriptCore/heap/HeapInlines.h:
```diff
@@ -78,11 +78,11 @@ ALWAYS_INLINE bool Heap::isMarked(const void* rawCell)
 ALWAYS_INLINE bool Heap::testAndSetMarked(HeapVersion markingVersion, const void* rawCell)
 {
-    HeapCell* cell = bitwise_cast<HeapCell*>(rawCell);
+    JSCell* cell = bitwise_cast<JSCell*>(rawCell);
```
This should still be `HeapCell` as the pointer may not be a JSCell.

kmiller68 commented on Source/JavaScriptCore/heap/MarkedBlock.cpp:
```diff
@@ -203,19 +204,22 @@ void MarkedBlock::Handle::resumeAllocating(FreeList& freeList)
 }
 
-void MarkedBlock::aboutToMarkSlow(HeapVersion markingVersion)
+void MarkedBlock::aboutToMarkSlow(HeapVersion markingVersion, JSCell* cell)
```
This should be `HeapCell` not `JSCell`

kmiller68 commented on Source/JavaScriptCore/heap/MarkedBlock.cpp:
```diff
@@ -502,6 +506,67 @@ void MarkedBlock::Handle::sweep(FreeList* freeList)
+        isCellVMSameAsBlockVM = blockVM == &cellVM;
+        if (isValidCellVM)
+            isAliveCell = VMInspector::isValidCell(&cellVM.heap, cell);
+    }
```
This isn't going to do anything since `HeapCell::vm()` just masks the bits off to get to the `MarkedBlock` and returns the `MarkedBlock`'s vm. Probably, the only meaningful parts of the cell, if it is one, is the cell's JSType and what offset into the MarkedBlock it is.

dhecht replied:
Rather than just the offset, it might be worth recording the entire address of the cell. We may find that it points into some region that doesn't make sense (but is mapped readable), meaning it wasn't the block header that was corrupted but that we chased an invalid pointer already.

hyjorc1 replied:
We should dump the addresses of the `VM` and `JSCell`. And the offset is just a subtraction.

kmiller68 commented on Source/JavaScriptCore/heap/MarkedBlock.h:
```diff
@@ -255,6 +257,8 @@ class MarkedBlock {
 
         static constexpr ptrdiff_t offsetOfVM() { return OBJECT_OFFSETOF(Header, m_vm); }
+
+        Handle* handleBitsForNullCheck() { return bitwise_cast<Handle*>(&m_handle); }
```
I don't get what this does? Why not just use `&handle()`?

kmiller68 commented on Source/JavaScriptCore/heap/MarkedBlock.h:
```diff
@@ -410,13 +414,16 @@ class MarkedBlock {
     inline bool marksConveyLivenessDuringMarking(HeapVersion markingVersion);
     inline bool marksConveyLivenessDuringMarking(HeapVersion myMarkingVersion, HeapVersion markingVersion);
+
+    // This is only used for debugging. We should remove this once the issue is resolved.
```
Link to radar to remove please.

dhecht commented on Source/JavaScriptCore/heap/MarkedBlock.cpp:
```diff
@@ -502,6 +506,67 @@ void MarkedBlock::Handle::sweep(FreeList* freeList)
+    VM* blockVM = header().m_vm;
+    {
+        isValidBlockVM = blockVM ? VMInspector::isValidVM(blockVM) : false;
+        if (isValidBlockVM) {
```
We know the pointer to handle in the header is corrupted (or we are already in some region that isn't actually a block), so it seems like there's a good chance m_vm field is also corrupted, and it may be corrupted with non-null values. So is there any value to having this 'then' path rather than just using the 'else' path always?

dhecht commented on Source/JavaScriptCore/heap/BlockDirectory.cpp:
```diff
@@ -388,6 +388,16 @@ void BlockDirectory::shrink()
+{
+    for (size_t index = 0; index < m_blocks.size(); ++index) {
+        MarkedBlock::Handle* handle = m_blocks[index];
+        if (handle->hasBlock(block))
```
rather than introduce hasBlock() (which should be removed later with all the other debugging code as it doesn't really make sense as a standalone interface), why not just do:

if (block == &handle->block())

MenloDorian replied:
Additionally, `hasBlock()` doesn't make sense because a Handle has a 1:1 correspondence with a block.  If we were to use a method here, a better name would be `isAssociatedWithBlock`.  `hasBlock` suggests that it can "contain" many blocks.

Anyway, I agree with Dan.  There's only one use of this right now, and a direct comparison makes sense here.

dhecht commented on Source/JavaScriptCore/heap/MarkedSpace.cpp:
```diff
@@ -358,6 +358,20 @@ bool MarkedSpace::isPagedOut()
     return pagedOutPagesStats.mean() > pagedOutPagesStats.count() * bailoutPercentage;
 }
 
+MarkedBlock::Handle* MarkedSpace::findMarkedBlockHandle(MarkedBlock* block)
```
let's not forget to remove these findMarkedBlockHandle methods as well once we're done with this debugging code

MenloDorian replied:
Let's rename this to `findMarkedBlockHandleDebug`.

MenloDorian commented on Source/JavaScriptCore/heap/BlockDirectory.cpp:
```diff
@@ -388,6 +388,16 @@ void BlockDirectory::shrink()
     }
 }
 
+MarkedBlock::Handle* BlockDirectory::findMarkedBlockHandle(MarkedBlock* block)
```
Let's call this `findMarkedBlockHandleDebug` to make it clear that we should not be calling this from normal code.

MenloDorian commented on Source/JavaScriptCore/heap/BlockDirectory.h:
```diff
@@ -146,7 +146,9 @@ class BlockDirectory {
     MarkedBlock::Handle* findBlockToSweep(unsigned& unsweptCursor);
-    
+
+    MarkedBlock::Handle* findMarkedBlockHandle(MarkedBlock*);
```
Ditto: let's name this appropriately.

MenloDorian commented on Source/JavaScriptCore/heap/MarkedBlock.h:
```diff
@@ -184,6 +184,8 @@ class MarkedBlock {
         bool isLiveCell(const void*);
 
+        bool hasBlock(MarkedBlock* block) { return m_block == block; }
+
```
Let's remove this.

MenloDorian commented on Source/JavaScriptCore/heap/MarkedBlock.cpp:
```diff
@@ -203,19 +204,22 @@ void MarkedBlock::Handle::resumeAllocating(FreeList& freeList)
-    
-    BlockDirectory* directory = handle().directory();
+
+    MarkedBlock::Handle* handle = header().handleBitsForNullCheck();
```
Let's name this `handlePointerForNullCheck` instead.

MenloDorian commented on Source/JavaScriptCore/heap/MarkedBlock.h:
```diff
@@ -255,6 +255,8 @@ class MarkedBlock {
 
         static constexpr ptrdiff_t offsetOfVM() { return OBJECT_OFFSETOF(Header, m_vm); }
+
+        Handle* handleBitsForNullCheck() { return &m_handle; }
```
Let's rename this.

MenloDorian commented on Source/JavaScriptCore/heap/MarkedBlock.h:
```diff
@@ -410,13 +412,16 @@ class MarkedBlock {
     inline bool marksConveyLivenessDuringMarking(HeapVersion markingVersion);
     inline bool marksConveyLivenessDuringMarking(HeapVersion myMarkingVersion, HeapVersion markingVersion);
+
+    // This is only used for debugging. We should remove this once the issue is resolved (https://rdar.apple.com/126213131)
```
https://rdar.apple.com/126213131 => rdar://126213131

MenloDorian commented on Source/JavaScriptCore/heap/MarkedBlock.h:
```diff
@@ -410,13 +412,16 @@ class MarkedBlock {
     inline bool marksConveyLivenessDuringMarking(HeapVersion myMarkingVersion, HeapVersion markingVersion);
+
+    // This is only used for debugging. We should remove this once the issue is resolved (https://rdar.apple.com/126213131)
+    NEVER_INLINE void dumpInfoIfHandleIsNotValid(AbstractLocker&, HeapCell*);
```
Let's call this `dumpInfoAndCrashForInvalidHandle` to make it clear that we're not returning.  You probably should add `NO_RETURN_DUE_TO_CRASH` here too.

MenloDorian commented on Source/JavaScriptCore/heap/MarkedSpace.h:
```diff
@@ -139,6 +139,8 @@ class MarkedSpace {
     void didConsumeFreeList(MarkedBlock::Handle*);
     void didAllocateInBlock(MarkedBlock::Handle*);
 
+    MarkedBlock::Handle* findMarkedBlockHandle(MarkedBlock*);
```
Let's rename this.

MenloDorian commented on Source/JavaScriptCore/heap/MarkedBlock.cpp:
```diff
@@ -502,6 +507,57 @@ void MarkedBlock::Handle::sweep(FreeList* freeList)
+        cellType,
+        contiguousZeroCount,
+        bitfield
+    );
```
Let's put this on a single line.

MenloDorian commented on Source/JavaScriptCore/heap/MarkedBlock.cpp:
```diff
@@ -502,6 +511,88 @@ void MarkedBlock::Handle::sweep(FreeList* freeList)
+    out.printf("found null handle in heap cell %p with type %d\n", heapCell, cellType);
+#if PLATFORM(COCOA)
+    WTF::setCrashLogMessage(out.toCString().data());
+#endif
```
Can you make and use a static template function for this instead?  e.g.
```
template<typename... Types>
static void logInvalidHandleDetails(const Types&... values)
{
    StringPrintStream out;
    out.printf(values...);
#if PLATFORM(COCOA)
    WTF::setCrashLogMessage(out.toCString().data());
#else
    dataLog(out.toCString().data());
#endif
}
```

Then, the code will be cleaner, and you won't forget to `WTF::setCrashLogMessage` it like you did below.

MenloDorian replied:
nit: Also, instead of printing 2 lines here, with this little info, there's no reason why this can't be a single line e.g.
```
    logInvalidHandleDetails("INVALID HANDLE: MarkedBlock %p HeapCell %p CellType %d\n", this, heapCell, cellType);
```
Note: I would prefix all these dumps with "INVALID HANDLE: ".  It gives us a handle to search for in the logs later (just in case it doesn't end up at the bottom for whatever reason).

kmiller68 commented on Source/JavaScriptCore/heap/MarkedBlock.h:
```diff
@@ -255,6 +255,14 @@ class MarkedBlock {
+        {
+            void* p = &m_handle;
+            Handle* target = nullptr;
+            memcpy(&target, &p, sizeof(uintptr_t));
```
I would just do `return opaque(&m_handle);`

kmiller68 commented on Source/JavaScriptCore/heap/MarkedBlock.cpp:
```diff
@@ -502,6 +511,86 @@ void MarkedBlock::Handle::sweep(FreeList* freeList)
+        });
+    }
+
+    uint64_t bitfield = 0xab00ab01ab020000;
```
What's up with this? Is it just to add bits for yourself to sanity check? I don't think it's necessary but not the end of the world.

danlliu replied:
Yes, just a sanity check to make sure we crashed at the end of the function.

kmiller68 replied:
For future reference, I don't think it's necessary since` WTFCrashWithInfo` passes a bunch of `__LINE__, __FILE__, WTF_PRETTY_FUNCTION, __COUNTER__` to make it invalid to deduplicate. But not worth worrying about here.

kmiller68 reviewed (approved):
r=me

MenloDorian commented on Source/JavaScriptCore/heap/MarkedBlock.cpp:
```diff
@@ -502,6 +511,86 @@ void MarkedBlock::Handle::sweep(FreeList* freeList)
+    if (!isBlockHandleInVM)
+        bitfield |= 1 << 5;
+
+    CRASH_WITH_INFO(cell, cellType, contiguousZeroCountAfterHandle, bitfield);
```
Let's add VM*, contiguousZeroCountAfterHandle as arguments as well.

MenloDorian reviewed (approved):
r=me

MenloDorian commented on Source/JavaScriptCore/heap/MarkedBlock.cpp:
```diff
@@ -502,6 +512,119 @@ void MarkedBlock::Handle::sweep(FreeList* freeList)
+        foundBlockInThisVM = isBlockInVM || isBlockHandleInVM;
+        LOG_INVALID_HANDLE_DETAILS("block in our VM = %d, block handle in our VM = %d\n", isBlockInVM, isBlockHandleInVM);
+
+        savedBitfield = (isValidBlockVM ? 8 : 0) | (isBlockInVM ? 4 : 0) | (isBlockHandleInVM ? 2 : 0) | (foundBlockInThisVM ? 1 : 0);
```
Can you SAVE_TO_REG savedBitfield here too?

MenloDorian commented on Source/JavaScriptCore/heap/MarkedBlock.cpp:
```diff
@@ -502,6 +512,119 @@ void MarkedBlock::Handle::sweep(FreeList* freeList)
+    bool isBlockHandleInVM = false;
+
+    VM* blockVM = header().m_vm;
+    VM* actualVM = header().m_vm;
```
This should be initialized to nullptr here, or remove actualVM completely because we don't know it yet.

MenloDorian commented on Source/JavaScriptCore/heap/MarkedBlock.cpp:
```diff
@@ -502,6 +512,119 @@ void MarkedBlock::Handle::sweep(FreeList* freeList)
+            if (blockVM == &vm) {
+                isValidBlockVM = true;
+                savedBitfield = 8;
+                LOG_INVALID_HANDLE_DETAILS("block VM %p is valid\n", &vm);
```
Let's SAVE_TO_REG savedActualVM here with &vm.

MenloDorian commented on Source/JavaScriptCore/heap/MarkedBlock.cpp:
```diff
@@ -502,6 +512,119 @@ void MarkedBlock::Handle::sweep(FreeList* freeList)
+    VM* blockVM = header().m_vm;
+    VM* actualVM = header().m_vm;
+    SAVE_TO_REG(savedBlockVM, "x23", blockVM);
+    SAVE_TO_REG(savedActualVM, "x24", actualVM);
```
This can just set savedActualVM to nullptr.

MenloDorian commented on Source/JavaScriptCore/heap/MarkedBlock.cpp:
```diff
@@ -502,6 +512,119 @@ void MarkedBlock::Handle::sweep(FreeList* freeList)
+                LOG_INVALID_HANDLE_DETAILS("block in another VM: %d, block in another VM: %d; other VM is %p\n", isBlockInVM, isBlockHandleInVM, &vm);
+
+                savedActualVM = actualVM;
+                savedBitfield = (isValidBlockVM ? 8 : 0) | (isBlockInVM ? 4 : 0) | (isBlockHandleInVM ? 2 : 0) | (foundBlockInThisVM ? 1 : 0);
```
To be consistent, can we also SAVE_TO_REG savedActualVM and savedBitfield here?  You can remove actualVM.  Don't think we need it.

kmiller68 reviewed (approved):
LGTM

danlliu committed:
Add some investigation code for dumping additional info when MarkedBlock::Handle is invalid
https://bugs.webkit.org/show_bug.cgi?id=280370
rdar://136782494

Reviewed by Mark Lam and Keith Miller.

We want to dump additional information when this failure case is hit to understand why this
could be taking place.

This patch is an updated version of Yijia's patch, which implemented most of the logging.

* Source/JavaScriptCore/heap/BlockDirectory.cpp:
(JSC::BlockDirectory::findMarkedBlockHandle):
* Source/JavaScriptCore/heap/BlockDirectory.h:
* Source/JavaScriptCore/heap/CellContainerInlines.h:
(JSC::CellContainer::aboutToMark):
* Source/JavaScriptCore/heap/HeapInlines.h:
(JSC::Heap::testAndSetMarked):
* Source/JavaScriptCore/heap/MarkedBlock.cpp:
(JSC::MarkedBlock::aboutToMarkSlow):
(JSC::MarkedBlock::dumpInfoIfHandleIsNotValid):
* Source/JavaScriptCore/heap/MarkedBlock.h:
(JSC::MarkedBlock::Handle::hasBlock):
(JSC::MarkedBlock::Header::handleBitsForNullCheck):
(JSC::MarkedBlock::aboutToMark):
* Source/JavaScriptCore/heap/MarkedSpace.cpp:
(JSC::MarkedSpace::findMarkedBlockHandle):
* Source/JavaScriptCore/heap/MarkedSpace.h:
* Source/JavaScriptCore/heap/SlotVisitorInlines.h:
(JSC::SlotVisitor::appendUnbarriered):
(JSC::SlotVisitor::appendHiddenUnbarriered):

Canonical link: https://commits.webkit.org/284376@main
